### PR TITLE
v4.3.0 — The Profiler, 25 Phantoms & a D-Bus Wire Protocol

### PR DESCRIPTION
**Pull request by johnzfitch**
# v4.3.0 &mdash; <mark>The Profiler, 25 Phantoms &amp; a D-Bus Wire Protocol</mark>

> [!NOTE]
> 34 commits. 1,980 lines added. Two days. We added a profiler flag, pointed it at ourselves, and found we&rsquo;d been software-rendering on an RTX 4090 while scanning 1,656 JSON files on every IPC call. We plugged 25 IPC handlers the webapp had been screaming into the void for. And to close [#28](https://github.com/johnzfitch/claude-cowork-linux/issues/28) &mdash; global shortcuts on Wayland &mdash; we wrote a D-Bus wire protocol client from scratch in 522 lines of Node.js, because Electron&rsquo;s `globalShortcut` uses X11&rsquo;s `XGrabKey`, and Wayland doesn&rsquo;t have an X11.[^xgrabkey]

[^xgrabkey]: `XGrabKey` on Wayland doesn&rsquo;t fail. It silently succeeds. It binds the key. The key never fires. Electron reports `true` from `register()`. You wait. Nothing happens. You add logging. The logging confirms the registration succeeded. You add more logging. Still nothing. Eventually you read the Electron source and find the X11 call and realise the entire abstraction is a polite fiction on Wayland compositors. The `xdg-desktop-portal` `GlobalShortcuts` interface is the real path &mdash; but it speaks D-Bus, and there&rsquo;s no Node.js library for it that doesn&rsquo;t pull in native bindings we can&rsquo;t ship.

---

## &starf; Three Workstreams

<dl>

<dt>1. IPC Surface Completeness &amp; Stability</dt>
<dd>

The webapp was calling 25 IPC handlers that didn&rsquo;t exist. Each call returned `undefined`. Each `undefined` cascaded into a validation failure, a greyed-out button, or a silent feature degradation that looked like a bug somewhere else entirely. The Startup preferences page was worse &mdash; it called `app.getLoginItemSettings()`, which is a macOS API, which throws on Linux, which crashes the main process. All 25 gaps filled, plus a plugin permission shim, an <abbr title="Model Context Protocol">MCP</abbr> reconnect storm fix, and a preload origin validation saga that took two attempts to get right.

</dd>

<dt>2. The Performance Blitz</dt>
<dd>

We added `./launch.sh --perf`. We ran it once. The profiler immediately highlighted `listLocalSessionMetadataFiles` &mdash; a recursive `readdirSync` that walked 1,656 JSON files and ran on **every single IPC round-trip** through the session read path. That&rsquo;s every `getSession`, every `getTranscript`, every `setFocusedSession`. It was a 1.2-second stall hiding behind a UI that looked merely &ldquo;sluggish.&rdquo; Six commits later: **&lt;1&nbsp;ms**. Also we discovered `--disable-gpu` was still on from early development. Software rendering. On a 4090.

</dd>

<dt>3. Platform Parity &amp; Wayland</dt>
<dd>

Closes **#28**. Electron&rsquo;s `globalShortcut` is X11-only, so we intercept it and route through `xdg-desktop-portal`&rsquo;s `GlobalShortcuts` D-Bus interface. The catch: no npm library speaks raw D-Bus without native bindings. So `portal_shortcuts.js` (522 lines) implements the wire protocol directly &mdash; alignment padding, variant types, `SIGNATURE` parsing, the works &mdash; shelling out to `gdbus` only for the socket path. Three wire-format bugs found and fixed before `CreateSession` succeeded and `hyprctl globalshortcuts` showed our keybinding. Also: user-agent spoofing, single-instance lock for Google OAuth, and a complete `install.sh` rewrite that stopped nuking the user&rsquo;s checkout.

</dd>
</dl>

---

## &starf; IPC Surface &mdash; The 25 Phantoms

<details>
<summary><strong>Handler Manifest</strong> (25 handlers across 5 namespaces)</summary>

#### Dispatch &amp; Bridge (12)

| Handler | Behavior |
|---------|----------|
| `getSessionsBridgeEnabled` | Returns persistent bridge state from closure |
| `setSessionsBridgeEnabled` | Stores bridge state &mdash; the fix for the vanishing toggle[^toggle] |
| `sessionsBridgeStatus` | Returns full status object with `enabled` flag |
| `getBridgeConsent` | Returns `{ consented: true }` &mdash; no consent gate on Linux |
| `abandonBridgeEnvironment` | No-op + trace |
| `resetBridge` | No-op + trace |
| `resetBridgeSession` | No-op + trace |
| `deleteBridgeSession` | No-op + trace |
| `deleteBridgeAgentMemory` | No-op + trace |
| `kickBridgePoll` | No-op + trace |
| `onBridgePermissionPreflight` | Listener registration shim |
| `respondBridgePermissionPreflight` | Auto-grants (no TCC on Linux) |

[^toggle]: The dispatch toggle disappeared after every tab navigation because `setSessionsBridgeEnabled` returned `undefined` and discarded the value. `getSessionsBridgeEnabled` had nothing to read back. A closure and a variable. That was the whole fix.

#### MCP, Permissions, Sessions, Plugins, Sharing (13)

| Category | Handler | Behavior |
|----------|---------|----------|
| <kbd>MCP</kbd> | `mcpCallTool` | Forward to MCP host |
| <kbd>MCP</kbd> | `mcpListResources` | Empty list |
| <kbd>MCP</kbd> | `mcpReadResource` | Empty list |
| <kbd>Perms</kbd> | `requestFolderTccAccess` | Auto-granted (Linux has no TCC) |
| <kbd>Perms</kbd> | `setChromePermissionMode` | No-op + trace |
| <kbd>Session</kbd> | `onCoworkFromMain` | Listener shim |
| <kbd>Session</kbd> | `onRemoteSessionStart` | Listener shim |
| <kbd>Session</kbd> | `openOutputsDir` | `xdg-open` on session output dir |
| <kbd>Session</kbd> | `setDraftSessionFolders` | Stores draft list |
| <kbd>Plugin</kbd> | `respondDirectoryServers` | Returns directory servers |
| <kbd>Plugin</kbd> | `respondPluginSearch` | Returns installed plugins |
| <kbd>Plugin</kbd> | `syncSkills` | Skills directory resync |
| <kbd>Share</kbd> | `shareSession` | No-op + trace (requires SSO) |

</details>

### The Plugin Permission Shim

The asar&rsquo;s marketplace flow expects a binary at a path determined by `process.resourcesPath` &mdash; the macOS `pluginShimPermissionBridge`. On Linux that binary doesn&rsquo;t exist. `cowork-plugin-shim.sh` is a 79-line shell script that speaks the same protocol: writes JSON permission requests to `shim-perm/requests/`, polls `shim-perm/responses/` for allow/deny decisions from the Desktop UI. Wired into both `launch.sh` (dev) and `PKGBUILD` (release).

### The Origin Validation Saga

The asar&rsquo;s `nue()` function validates that IPC senders have an `https://claude.ai` origin. Running via `electron .asar` sets `process.defaultApp=true`, making `app.isPackaged=false`. The validator rejects `file://` URLs when `isPackaged` is false. Result: the preload&rsquo;s `getInitialLocale()` throws before `contextBridge.exposeInMainWorld` runs. Every renderer gets `process is not defined`.

**Attempt 1** (`e1a0657`): Wrap `webContents.ipc.on()` to proxy `senderFrame` for `DesktopIntl` channels, making validation see `https://claude.ai` for `file://` preloads. **Problem**: timing race &mdash; the asar registers `DesktopIntl` handlers *before* `web-contents-created` fires, so our wrapper never intercepts.

**Attempt 2** (`e4a2099`): `sed`-patch the asar at repack time:

```bash
# BEFORE:  e.protocol==="file:"&&Ee.app.isPackaged===!0
# AFTER:   e.protocol==="file:"
```

One line. The `file://` content is inside our asar. No security implications. Removed the dead `ipc.on()` wrapping.

> [!CAUTION]
> The Electron module patching code (`Module.prototype.require('electron')` interception) was running on **every** `require()` call across the entire process, adding duplicate `web-contents-created` listeners each time. A `MaxListenersExceededWarning` at 11 listeners confirmed the leak. `global.__coworkElectronPatched` gates it to run exactly once. If you ever see that warning come back, check here first.

<details>
<summary><strong>Remaining Stability Commits</strong></summary>

| Commit | Fix |
|--------|-----|
| `d6054f0` | **Phase 1 PR review fixes**: remove OAuth token value from trace log (log only presence + length), fix `forwardEvent` bridge patch delegating to `originalForwardEvent` for non-message events, remove Atomics.wait retry loop that blocked main process ~400ms on non-dispatch spawns |
| `c752206` | **Bridge diagnostics**: check `transport.closed` before writing, log successful writes with message type and session ID |
| `6ca9545` | **Auth rejection logging**: empty or CRLF-injected tokens in `buildAuthHeaders` were silently swallowed. Now `console.warn` with rejection reason. |
| `882a6de` | **Disclaimer overwrite**: `existsSync` guard prevented the platform-aware disclaimer from replacing the old passthrough. The macOS binary path redirect never activated, causing exit code 126 on marketplace operations. |

</details>

---

## &starf; The Performance Blitz

`./launch.sh --perf` is new this PR. It hooks up the Node.js inspector on port 9229, auto-opens DevTools, and sets `CLAUDE_COWORK_VERBOSE=1` for diagnostic logging. The first profiling session took about ten seconds to find the problem:

```
listLocalSessionMetadataFiles()
  -> recursive readdirSync  -> 1,656 JSON files
  -> called on EVERY IPC round-trip through:
       observeSessionRead()
         -> normalizeSessionRecord()
           -> findSessionMetadataPath()
```

The app is fundamentally a session manager. Every user interaction is a session-related IPC call. Every one of those calls was doing a full recursive directory walk, reading and parsing JSON, and searching for transcript candidates. **Every. Time.**

<details>
<summary><strong>The Fix Chain</strong> (6 performance commits + 2 logging commits)</summary>

<table>
<thead>
<tr><th>Commit</th><th>What</th><th>Impact</th></tr>
</thead>
<tbody>
<tr>
<td><code>8aa2417</code></td>
<td>Drop <code>--disable-gpu</code>, enable hardware acceleration</td>
<td>The flag was a &ldquo;safety measure&rdquo; from early development. Software rendering on an RTX 4090 is not safe. It&rsquo;s slow. <code>nvidia-drm</code> is stable on Wayland/Hyprland. The flag is gone.</td>
</tr>
<tr>
<td><code>b96b02a</code></td>
<td>Cache <code>listLocalSessionMetadataFiles</code> with 5s TTL + O(1) basename&rarr;path index</td>
<td><strong>1.2s &rarr; &lt;1ms</strong> per IPC call. The cache is a <code>Map</code> with basename keys for constant-time lookup. <code>invalidateMetadataFileCache()</code> exported for session create/delete.</td>
</tr>
<tr>
<td><code>145aaa7</code></td>
<td>Defer session metadata normalization to access time</td>
<td>Session list loads no longer eagerly normalize every session record upfront</td>
</tr>
<tr>
<td><code>2a83bb4</code></td>
<td>Cache <code>normalizeSessionRecord</code> + <code>getSessionInfo</code> with 2s TTL</td>
<td>All IPC calls within a single user action (getSession, getTranscript, setFocused) share one cached result &mdash; no repeated filesystem reads or JSON parsing</td>
</tr>
<tr>
<td><code>91d41ec</code></td>
<td>Remove dead <code>~/Library/Application Support/Claude/claude-code-vm</code> from <code>claudeVmRoots</code></td>
<td>Eliminates a <code>stat()</code> call per path resolution (macOS path that never exists on Linux)</td>
</tr>
<tr>
<td><code>b8e52fe</code></td>
<td>Fast-path <code>redactCredentials</code> when no token prefixes present</td>
<td>Single regex pre-check before running 4 replacement passes + Shannon entropy. <strong>Also fixes false-positive redaction of long <code>PATH</code> values</strong> (high character diversity triggered the entropy heuristic)</td>
</tr>
<tr>
<td><code>bb82113</code></td>
<td>Remove unconditional <code>new Error().stack</code> on Swift stub load</td>
<td>Stack trace generation is expensive in V8 and served no purpose outside initial debugging</td>
</tr>
<tr>
<td><code>15e6116</code></td>
<td>Gate ~30 <code>console.log</code> calls behind <code>CLAUDE_COWORK_VERBOSE=1</code></td>
<td>Production launches emit only errors and critical state changes. <code>--perf</code> sets the flag for diagnostic sessions.</td>
</tr>
</tbody>
</table>

</details>

---

### The D-Bus Wire Protocol &mdash; Closing [#28](https://github.com/johnzfitch/claude-cowork-linux/issues/28)

Electron&rsquo;s `globalShortcut.register()` calls `XGrabKey`. On Wayland that silently does nothing. The correct path is `org.freedesktop.portal.GlobalShortcuts` &mdash; a D-Bus interface that every major Wayland compositor implements. The catch: every Node.js D-Bus library pulls in native bindings (`dbus-native`, `dbus-next`, `node-dbus`) that we can&rsquo;t ship inside an Electron app without a build step.

`portal_shortcuts.js` (522 lines) implements the D-Bus wire protocol from scratch. Pure Node.js. Zero npm dependencies. Uses `gdbus` (ships with glib2) only to discover the socket path, then speaks raw binary over a Unix domain socket:

```
 +------------------------------------------------------+
 |  frame-fix-wrapper.js                                |
 |    intercepts electron.globalShortcut.register()     |
 |    -> isWayland? route to portal : use Electron native|
 +----------------------+-------------------------------+
                        |
 +----------------------v-------------------------------+
 |  portal_shortcuts.js                                 |
 |    D-Bus wire protocol: alignment, variants, sigs    |
 |    CreateSession -> BindShortcuts -> Activated signal |
 |    -> callback fires on keypress                     |
 +----------------------+-------------------------------+
 |                      | raw bytes over AF_UNIX
 +----------------------v-------------------------------+
 |  xdg-desktop-portal (compositor)                     |
 |    org.freedesktop.portal.GlobalShortcuts             |
 +------------------------------------------------------+
```

Three wire-format bugs had to be fixed before `CreateSession` succeeded (`e69d962`):

1. **Alignment**: `bWriteStr`/`bWriteVariantStr` didn&rsquo;t enforce 4-byte alignment for string values, causing `UINT32` length fields to land at wrong offsets inside variants
2. **Array length**: `bWriteDictSV` included alignment padding between the `UINT32` length field and the first dict entry in the byte count &mdash; the D-Bus spec excludes that padding
3. **SIGNATURE type**: `readVariant` didn&rsquo;t handle `g` (SIGNATURE), so when the portal sent header fields with SIGNATURE before MEMBER (field 8 before field 3), the parser got stuck and all subsequent fields parsed as empty

After the fix: `hyprctl globalshortcuts` shows the `claude-ctrl-alt-space` entry. Works on KDE, Hyprland, GNOME&nbsp;48+. Falls back to Electron&rsquo;s native `XGrabKey` on X11 or when the portal is absent.

> [!TIP]
> The asar defaults to `Alt+Space` because we spoof `darwin`. On Linux, `Alt+Space` is the window menu in GNOME, the app launcher in Hyprland, and the search bar in KDE. We write `Ctrl+Alt+Space` into `config.json` on first run if the user hasn&rsquo;t set a custom shortcut.[^shortcut]

[^shortcut]: `Ctrl+Alt+Space` is not sacred either. It just doesn&rsquo;t collide with any default keybinding on the three major compositors. The user can override it in Settings &rarr; General &rarr; Keyboard Shortcut.

### The Consequences of Spoofing Darwin

We spoof `process.platform` to `darwin` because the asar&rsquo;s code paths assume macOS. Each spoof creates a downstream consequence that needs its own fix:

| Consequence | Fix |
|-------------|-----|
| Asar constructs macOS binary paths (`claude.app/Contents/MacOS/claude`), tries to exec an arm64 Mach-O on x86 Linux (exit 126) | Platform-aware disclaimer wrapper resolves to the Linux binary (`5a689cb`) |
| `Alt+Space` shortcut default conflicts with every WM launcher | Write `Ctrl+Alt+Space` on first run (`766891e`) |
| Webapp&rsquo;s UA check sees `Linux`, falls through to &ldquo;Claude for Windows&rdquo; branding | Spoof UA to `Macintosh; Intel Mac OS X 10_15_7` (`8b97603`) |
| macOS `open-url` codepath expects Apple Event routing for `claude://` protocol | Single-instance lock, forward via `open-url` event (`8e129f6`) |
| `os.platform()` / `os.arch()` overrides referenced deleted `_inOurCode` flag | Clean up stale refs (`9ef5698`, `0a2b763`) |

### Install &amp; Launch UX

<table>
<thead>
<tr><th>Commit</th><th>Before</th><th>After</th></tr>
</thead>
<tbody>
<tr>
<td><code>cea618a</code></td>
<td><code>rm -rf linux-app-extracted/</code> then extract</td>
<td>Extract over existing tree &mdash; overwrites stale files, preserves extras</td>
</tr>
<tr>
<td><code>9ef5698</code></td>
<td>Stale files from older asars persist after re-extraction</td>
<td>Extract to temp dir then atomic swap. Preserves <code>resources/</code> (i18n, icons from DMG overlay).</td>
</tr>
<tr>
<td><code>5a5a543</code></td>
<td>Prompt: <em>type <kbd>yes</kbd> to continue</em></td>
<td>Log and proceed</td>
</tr>
<tr>
<td><code>bf3a291</code></td>
<td>Sync stubs from install dir (stale if <code>git pull</code> failed)</td>
<td>Sync stubs from the repo dir where <code>install.sh</code> was invoked</td>
</tr>
<tr>
<td><code>ca3804b</code></td>
<td><code>fetch-dmg.js</code> from install dir hits <code>claude.ai/api/desktop</code> (now Cloudflare 403)</td>
<td><code>fetch-dmg.js</code> from repo dir uses Homebrew cask API</td>
</tr>
<tr>
<td><code>b140bf9</code></td>
<td><code>./launch.sh</code> blocks the terminal</td>
<td>Backgrounds Electron, returns the prompt. <code>--dev</code>/<code>--devtools</code>/<code>--perf</code> for foreground.</td>
</tr>
<tr>
<td><code>d762716</code></td>
<td><code>install.sh</code> launcher stripped <code>--devtools</code>/<code>--debug</code> before exec</td>
<td>Passes flags through so foreground mode detection still triggers</td>
</tr>
</tbody>
</table>

---

## &starf; Test Coverage

<samp>376 tests</samp> &ensp;&bull;&ensp; <samp>372 pass</samp> &ensp;&bull;&ensp; <samp>4 pre-existing</samp>[^conformance]

[^conformance]: The 4 failures are in `webapp_conformance.test.cjs` &mdash; `vmDownloadStatus` and `vmRunningStatus` expected values changed in a newer webapp dump. They predate this branch and are unrelated to our work.

| Suite | Delta | Coverage |
|-------|-------|---------|
| `ipc_overrides.test.cjs` | <mark>new file, +261 lines</mark> | All 25 IPC handlers, bridge state persistence, Startup shims, MCP debounce, disclaimer wrapper |
| `portal_shortcuts.test.cjs` | <mark>new file, +189 lines</mark> | D-Bus wire format: alignment, variant encoding, `CreateSession`/`BindShortcuts` round-trip, signal parsing, fallback to Electron native |
| `session_orchestrator.test.cjs` | +70 lines | Normalization delegation, live event path, bridge env injection |
| `transcript_store.test.cjs` | +31 lines | Metadata edge cases |
| `sessions_api.test.cjs` | +19 lines | Auth header rejection logging |
| `credential_classifier.test.cjs` | +11 lines | Fast-path validation, `PATH` false-positive regression |
| `session_store.test.cjs` | +9 lines | Cache invalidation |
| `audit_regression.test.cjs` | +5 lines | Regression guards for Copilot audit fixes |

---

## &starf; What&rsquo;s Next

| &numero; | Issue | Status |
|------|-------|--------|
| [#58](https://github.com/johnzfitch/claude-cowork-linux/issues/58) | <abbr title="Single Sign-On">SSO</abbr> and Google sign-in | Partial &mdash; UA spoof fixes Google OAuth&rsquo;s platform check. The cookie-based flow (`ASWebAuthenticationSession`) still routes to `shell.openExternal`, which loses the session cookie. Workaround: email/password login. |
| [#52](https://github.com/johnzfitch/claude-cowork-linux/issues/52) | Code panel send button greyed out | Needs investigation into the asar&rsquo;s code panel activation logic. |
| Dispatch end-to-end | `forwardEvent` patch installed, diagnostics in place | The patch fires for `user` events. `result`/`stream_event` types aren&rsquo;t reaching it. The `--perf` diagnostics will show exactly where. |

> [!WARNING]
> `--remote-debugging-port=9222` in the `--perf` block is for development only &mdash; it&rsquo;s how `chrome-devtools-mcp` connects to the Electron renderer. Any local process can read renderer memory and inject code through this port. Remove before any public release build.

---

<details>
<summary><strong>Full Commit Log</strong> (34 commits, reverse chronological)</summary>

| SHA | Message |
|-----|---------|
| `0a2b763` | fix: remove stale _inOurCode refs from os.platform/os.arch overrides |
| `9ef5698` | fix: asar extraction stale files, remove dead withRealPlatform |
| `d762716` | fix: PR review &mdash; launcher foreground mode, portal unbind, register fallback |
| `e69d962` | fix: D-Bus wire protocol bugs in portal shortcuts client |
| `54b8bf3` | feat: global shortcuts via xdg-desktop-portal on Wayland |
| `766891e` | fix: default global shortcut to Ctrl+Alt+Space on Linux |
| `cea618a` | fix: install.sh overlays instead of nuking directories |
| `bf3a291` | fix: install.sh uses stubs from script dir, not stale install dir |
| `8e129f6` | fix: single-instance lock for Google auth protocol callback |
| `8b97603` | fix: spoof user agent to macOS for webapp platform detection |
| `b140bf9` | fix: default launch.sh to headless, --dev for foreground terminal |
| `5a5a543` | fix: remove scary rm-rf confirmation prompt from install.sh |
| `ca3804b` | fix: prefer script-dir fetch-dmg.js over stale install-dir copy |
| `15e6116` | fix: gate all verbose logging behind --perf (CLAUDE_COWORK_VERBOSE) |
| `bb82113` | fix: remove unconditional stack trace and verbose logging from swift stub |
| `6f1e5c1` | feat: add --remote-debugging-port to --perf for CDP MCP access |
| `b8e52fe` | perf: fast-path redactCredentials when no sensitive patterns present |
| `91d41ec` | perf: remove dead macOS legacy path from claudeVmRoots |
| `b599739` | fix: guard electron module patching against re-entry |
| `2a83bb4` | perf: cache normalizeSessionRecord and getSessionInfo (2s TTL) |
| `145aaa7` | perf: defer session metadata normalization to access time |
| `b96b02a` | perf: cache session metadata file listing (1.2s &rarr; &lt;1ms) |
| `a90a188` | feat: add --perf flag for CPU profiling |
| `f4860a3` | perf: eliminate stack trace generation from platform spoofing |
| `8aa2417` | perf: drop --disable-gpu, enable hardware acceleration |
| `e4a2099` | fix: patch asar origin validation for file:// preloads |
| `e1a0657` | fix: preload origin validation, Startup IPC, MCP reconnect storm |
| `882a6de` | fix: always overwrite disclaimer wrapper on startup |
| `6ad816b` | fix: persist dispatch bridge enabled state across tab navigations |
| `6ca9545` | fix: log when buildAuthHeaders rejects a token |
| `5a689cb` | fix: shim ENOENT, disclaimer exec format, Startup handlers |
| `9fcb316` | feat: plugin shim + 25 missing IPC handlers |
| `c752206` | fix: add transport.write diagnostics to bridge patch |
| `d6054f0` | fix: Phase 1 PR review fixes from Copilot audit |

</details>

---

<sub>30 files changed &ensp;&bull;&ensp; +1,980 &minus;302 &ensp;&bull;&ensp; 376 tests &ensp;&bull;&ensp; 372 pass, 4 pre-existing &ensp;&bull;&ensp; closes <a href="https://github.com/johnzfitch/claude-cowork-linux/issues/28">#28</a></sub>

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -471,8 +494,8 @@ mkdir -p "\$LOG_DIR"
 cd "\$COWORK_DIR"
 
 case "\${1:-}" in
-    --devtools) shift; export CLAUDE_DEVTOOLS=1; exec ./launch.sh "\$@" 2>&1 | tee -a "\$LOG_DIR/startup.log" ;;
-    --debug)    shift; export CLAUDE_TRACE=1; exec ./launch.sh "\$@" 2>&1 | tee -a "\$LOG_DIR/startup.log" ;;
+    --devtools) shift; export CLAUDE_DEVTOOLS=1; exec ./launch.sh "\$@" ;;
```
Comment (johnzfitch) [inline install.sh:497]: `claude-desktop --devtools` and `claude-desktop --debug` no longer keep the process in the foreground. The generated launcher strips the flag before calling `./launch.sh`, but `launch.sh` only sets `_dev=true` when it sees `--dev`, `--devtools`, or `--perf` in argv. After this change both code paths here just export an env var and call `./launch.sh` with no dev flag, so it falls into the detached `nohup` branch and immediately backgrounds. That breaks the advertised debugging workflow because the terminal never stays attached and the user loses the live logs/devtools behavior promised by the help text below.

File: install.sh [hunk 2]
```diff
@@ -345,11 +356,7 @@ extract_archive() {
     local asar_file="$claude_app/Contents/Resources/app.asar"
     [[ -f "$asar_file" ]] || die "app.asar not found at: $asar_file"
 
-    # Extract asar into linux-app-extracted/
-    if [[ -d "$target_dir" ]]; then
-        log_info "Removing previous linux-app-extracted..."
-        rm -rf "$target_dir"
-    fi
+    # Extract on top of existing tree (overwrites stale files, preserves extras)
```
Comment (johnzfitch) [inline install.sh:359]: Extracting on top of the old tree means files that were removed from the upstream `app.asar` never get deleted locally, and the next `asar pack` will ship those stale files forever. The previous version explicitly removed `linux-app-extracted` first, which kept the repacked tree in sync with the current Claude build. With this change, a rename/delete in a future upstream release can leave dead JS/assets/native files in the repacked app and produce version-skew bugs that are very hard to diagnose.
